### PR TITLE
Integration of SPDX License Compliance Checker workflows

### PR DESCRIPTION
**Pull request by anirudTT**
#### Overview

This pull request introduces changes to GitHub Actions workflows, focusing on ensuring adherence to SPDX license requirements for Python files. It adds two new workflows, `check-license-year.yml` and `license-checker.yml`, to automate the verification of SPDX license headers and the checking of the correct license year within all _changed_ and _new_ Python files involved/added via a PR to `main`.

#### Additions / Changes to repository via PR or otherwise :

1. **SPDX License Year Validation Workflow (`check-license-year.yml`)**:
   - Automates the validation of the SPDX license year within changed Python files in a PR against the current year (2024).
   - Flags files missing the correct year or without an SPDX license identifier and adds a comment on PR listing those flagged files , and prevents PR merging until compliance is achieved.

2. **SPDX License Header Check Workflow (`license-checker.yml`)**:
   - Checks for the presence of valid SPDX license headers in Python files included in the PR, according to predefined rules in `check_copyright_config.yaml`.
   - Lists non-compliant files in an automated PR comment, requesting updates to meet the project's licensing standards.

3. **Configuration File (`check_copyright_config.yaml`)**:
   - Defines the criteria for the SPDX license checks, including allowed licenses (Apache-2.0, MIT), the default license for new files, and a custom notice template for Python files.

4. **Updates to CONTRIBUTING.md**:: to better explain contributors on how to navigate workflow failures and passing criteria. 

5. **Outside the PR**: https://github.com/tenstorrent/benchmarking/settings/rules --- rule set will be  added to prevent a PR being merged unless all checks are passed 

#### How It Works

Upon opening or updating a PR, the newly implemented workflows run automatically. Non-compliance with our SPDX licensing standards triggers informative comments within the PR, illustrating precisely what needs to be corrected. (See included screenshots for examples of these automated comments.)

#### Screenshots

1. PR comments:
Screenshot 2024-04-10 at 12 22 28 PM

2. Shows clearly that the workflow failed:
Screenshot 2024-04-10 at 12 22 41 PM

### CODE REVIEW

File: check_copyright_config.yaml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+  # List of allowed licenses for files in this section
+  allowed_licenses:
+    - Apache-2.0
+    - MIT
+  license_for_new_files: Apache-2.0  # License to be used when inserting a new copyright notice
+
+  new_notice_python: |
+    """
+    SPDX-FileCopyrightText: © {years} Tenstorrent Inc.
```
Comment (milank94) [inline check_copyright_config.yaml:12]:
```suggestion
    SPDX-FileCopyrightText: © {years} Tenstorrent AI ULC
```

File: check_copyright_config.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+    SPDX-License-Identifier: {license}
+    """
+
+  # Define the template for replacing comment lines with SPDX notation
+  espressif_copyright: '© {years} Tenstorrent Inc.'
```
Comment (milank94) [inline check_copyright_config.yaml:18]:
```suggestion
  espressif_copyright: '© {years} Tenstorrent AI ULC'
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -98,11 +96,45 @@ or more information on the GitHub Actions and Pull Request Workflow, please see
...
+The GitHub Actions workflow also automatically makes a commit with the message `*** AUTOMATED COMMIT | Applied Code Formatting and Cleanup ✨ 🍰 ✨***` authored by `[anirudTT]` when it performs code formatting and cleanup. If you open a pull request and subsequently push more changes, we suggest **rebasing or pulling again** from the pull request branch before pushing your changes again to avoid conflicts.
+
+#### SPDX License Compliance Check
+
+Our project enforces SPDX license compliance for all Python files through automated checks integrated within our GitHub Actions workflows. These workflows, `check-license-year.yml` and `license-checker.yml`, automatically verify the presence and accuracy of SPDX license headers and the specified license year in Python files included in pull requests.
```
Comment (milank94) [inline CONTRIBUTING.md:108]: We should be doing this for all file types, not just Python i.e. C, Bash, etc. I know we don't have any yet, but we may in the future. Maybe just remove the strictness on Python for now.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -98,11 +96,45 @@ or more information on the GitHub Actions and Pull Request Workflow, please see
...
+
+- **License Year Validation**: The `check-license-year.yml` workflow checks for the correct SPDX license year (`2024`) in newly changed Python files. If files lacking the correct year or license identifier are detected, the workflow comments on the PR with details on the missing elements and fails to highlight the need for compliance.
```
Comment (milank94) [inline CONTRIBUTING.md:110]: Anyway for us to automate the year it verifies against? Somehow extract the current year from a system time.

This way we don't have to worry about updating the repo every new year.

File: CONTRIBUTING.md (hunk 3 above)
Comment (anirudTT) [reply CONTRIBUTING.md]: Commit https://github.com/tenstorrent/benchmarking/pull/17#:~:text=Resolve%20conversation-,Test%20dynamic%20year%20(,-%2311) fixes the mentions of just python files .

**Review by anirudTT — COMMENTED** [review]
made requested changes

File: .github/workflows/check-license.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        if: env.missing_license_files != ''
+        uses: actions/github-script@v6
+        with:
+          script: |
+            const issueNumber = context.issue.number;
+            const owner = context.repo.owner;
+            const repo = context.repo.repo;
+            const missingFiles = process.env.missing_license_files.trim().replace(/\s+/g, '\n');
+            const currentYear = process.env.current_year;
+            const commentBody = `The following Python files are not compliant with the required licensing standards for ${currentYear} or do not contain the correct copyright text 'Tenstorrent AI ULC' within their license header:\n\`\`\`\n${missingFiles}\n\`\`\`\nPlease update the files accordingly. Your attention and cooperation in this matter are greatly appreciated.\nThank you.`;
```
Comment (milank94) [inline .github/workflows/check-license.yml:47]: Should we remove "Python files" from here and just say "files"?

Would this work with other file formats i.e. .c, .cpp, etc.?

I think we would just need to add the extension to `check_copyright_config.yaml` but then it should check all of the formats?

Comment (anirudTT) [reply .github/workflows/check-license.yml]: > Should we remove "Python files" from here and just say "files"?

I just updated it to just say files ; That's a good catch. 

> Would this work with other file formats i.e. .c, .cpp, etc.?

Currently i dont think so , `license-checker.yml` uses the `check_copyright_config.yaml` which based on [this ](https://github.com/anirudTT/benchmarking/blob/main/check_copyright_config.yaml#L8) should check on other files but frankly I need to test this; as I have only been testing python files.

> I think we would just need to add the extension to check_copyright_config.yaml but then it should check all of the formats?

Yes ; to check other files ; but I think currently let's keep it this way and based on whats gets added update the config , what do you think?

**Review by anirudTT — COMMENTED** [review]
Updated:

1. the workflow name as I felt both almost are named the same 
2. updated to remove the name python files from PR comment body

File: .github/workflows/check-license.yml (hunk 5 above)
Comment (milank94) [reply .github/workflows/check-license.yml]: Exactly -- no action; this was more out of my curiosity. When a different file format comes in we will test that.

**Review by milank94 — APPROVED** [review]
Very nice! Looks good

File: .github/workflows/check-license.yml (hunk 5 above)
Comment (anirudTT) [reply .github/workflows/check-license.yml]: I think let me also test this on my dev branch -- just to figure out how its going to act

**Pull request merged**